vault/identity_store_util.go

```diff
@@ -0,0 +1,1244 @@
+package vault
+
+import (
+	"fmt"
+	"strings"
+	"sync"
+	"time"
+
+	memdb "github.com/hashicorp/go-memdb"
+	uuid "github.com/hashicorp/go-uuid"
+	"github.com/hashicorp/vault/helper/consts"
+	"github.com/hashicorp/vault/helper/locksutil"
+	"github.com/hashicorp/vault/logical"
+)
+
+func (c *Core) loadEntities() error {
+	if c.identityStore == nil {
+		return fmt.Errorf("identity store is not setup")
+	}
+
+	return c.identityStore.loadEntities()
+}
+
+func (i *identityStore) loadEntity(entity *entityStorageEntry, previousEntity *entityStorageEntry) error {
+	var err error
+
+	if entity == nil {
+		return fmt.Errorf("entity is nil")
+	}
+
+	// Acquire the lock to modify the entity storage entry
+	lock := locksutil.LockForKey(i.entityLocks, entity.ID)
+	lock.Lock()
+	defer lock.Unlock()
+
+	// Create a MemDB transaction to update both identity and entity
+	txn := i.db.Txn(true)
+	defer txn.Abort()
+
+	for _, identity := range entity.Identities {
```

Comment:
This is a very good question. IIUC, what you mean to say is that when this function is run, there is no scope for concurrency anyways since Vault will be sealed, and why not avoid locking altogether. Sgtm. I'll remove them.